### PR TITLE
Initial transaction support

### PR DESCRIPTION
**Pull request by richard-rogers**
## Description

Support for transactions
```
    writer = WhyLabsWriter(dataset_id=MODEL_ID)
    writer.start_transaction()
    status, id = writer.write(result)
    ...
    writer.commit_transaction()
```

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: python/whylogs/api/writer/whylabs.py [hunk 1]
```diff
@@ -11,21 +11,29 @@
 from whylabs_client import ApiClient, Configuration  # type: ignore
 from whylabs_client.api.dataset_profile_api import DatasetProfileApi  # type: ignore
 from whylabs_client.api.feature_weights_api import FeatureWeightsApi  # type: ignore
-from whylabs_client.api.log_api import AsyncLogResponse  # type: ignore
+
+# from whylabs_client.api.log_api import AsyncLogResponse  # type: ignore
```
Comment (FelipeAdachi) [inline python/whylogs/api/writer/whylabs.py]: remove the commented line?

File: python/whylogs/api/writer/whylabs.py [hunk 2]
```diff
@@ -606,6 +615,35 @@ def _tag_custom_perf_metrics(self, view: Union[DatasetProfileView, SegmentedData
...
+        self._transaction_id = result["transaction_id"]
+
+    def commit_transaction(self, **kwargs) -> None:
+        if self._transaction_id is None:
+            logger.error("Must call start_transaction() before commit_transaction()")
+            return
+
+        client = self._get_or_create_transaction_client()
+        request = TransactionCommitRequest(verbose=True)
+        print(f"comminting {self._transaction_id}")
```
Comment (FelipeAdachi) [inline python/whylogs/api/writer/whylabs.py]: typo "comminting"

File: python/whylogs/api/writer/whylabs.py [hunk 3]
```diff
@@ -606,6 +615,35 @@ def _tag_custom_perf_metrics(self, view: Union[DatasetProfileView, SegmentedData
...
+        result: Response = client.commit_transaction(self._transaction_id, request, **kwargs)
+        print(result)
```
Comment (FelipeAdachi) [inline python/whylogs/api/writer/whylabs.py]: are these prints just for debugging? if so, let's remove them

File: python/whylogs/api/writer/whylabs.py [hunk 4]
```diff
@@ -679,10 +717,26 @@ def write(self, file: Writable, **kwargs: Any) -> Tuple[bool, str]:
                 )
 
             dataset_timestamp_epoch = int(stamp * 1000)
-            response = self._do_upload(
-                dataset_timestamp=dataset_timestamp_epoch,
-                profile_file=tmp_file,
-            )
+            if self._transaction_id is not None:
```
Comment (FelipeAdachi) [inline python/whylogs/api/writer/whylabs.py:726]: we have different codepaths for feature weights, estimation result, or segmented result sets. If I'm understanding this correctly, transactional support won't work for these cases, right?

File: python/whylogs/api/writer/whylabs.py (hunk 1 above)
Comment (richard-rogers) [inline python/whylogs/api/writer/whylabs.py]:
```suggestion
```

File: python/whylogs/api/writer/whylabs.py [hunk 5]
```diff
@@ -606,6 +614,35 @@ def _tag_custom_perf_metrics(self, view: Union[DatasetProfileView, SegmentedData
...
+        self._transaction_id = result["transaction_id"]
+
+    def commit_transaction(self, **kwargs) -> None:
+        if self._transaction_id is None:
+            logger.error("Must call start_transaction() before commit_transaction()")
+            return
+
+        client = self._get_or_create_transaction_client()
+        request = TransactionCommitRequest(verbose=True)
+        print(f"comminting {self._transaction_id}")
```
Comment (richard-rogers) [inline python/whylogs/api/writer/whylabs.py]:
```suggestion
```

File: python/whylogs/api/writer/whylabs.py [hunk 6]
```diff
@@ -606,6 +614,34 @@ def _tag_custom_perf_metrics(self, view: Union[DatasetProfileView, SegmentedData
...
+
+    def commit_transaction(self, **kwargs) -> None:
+        if self._transaction_id is None:
+            logger.error("Must call start_transaction() before commit_transaction()")
+            return
+
+        client = self._get_or_create_transaction_client()
+        request = TransactionCommitRequest(verbose=True)
+        result: Response = client.commit_transaction(self._transaction_id, request, **kwargs)
+        print(result)
```
Comment (richard-rogers) [inline python/whylogs/api/writer/whylabs.py]:
```suggestion
```

File: python/whylogs/api/writer/whylabs.py (hunk 4 above)
Comment (richard-rogers) [reply python/whylogs/api/writer/whylabs.py:726]: Segmented profiles call back to `write()` for each segment, so I think those would work. I think segmented _reference_ profiles and the other ones you mention would need additional transactional client API endpoints, or perhaps make transaction ID an optional parameter on those endpoints... Right now, writing those things when a transaction is open works, but they're not included in the transaction. Another option would be to issue a warning or error on writing non-transactable things when a transaction is open. Or maybe have `start_transaction()` return an ID that gets passed to `write()` as an optional keyword argument. I'd prefer not do add more transaction-specific methods to the `WhyLabsWriter` without a major refactoring.

Comment (FelipeAdachi) [reply python/whylogs/api/writer/whylabs.py:726]: other than seg reference profiles, the other ones are very lightweight, so maybe for now we could just emit a warning for when they are being uploaded while a transaction is open?

File: python/whylogs/api/writer/whylabs.py [hunk 7]
```diff
@@ -606,6 +612,33 @@ def _tag_custom_perf_metrics(self, view: Union[DatasetProfileView, SegmentedData
...
+        self._transaction_id = result["transaction_id"]
+
+    def commit_transaction(self, **kwargs) -> None:
+        if self._transaction_id is None:
+            logger.error("Must call start_transaction() before commit_transaction()")
+            return
+
+        client = self._get_or_create_transaction_client()
+        request = TransactionCommitRequest(verbose=True)
+        client.commit_transaction(self._transaction_id, request, **kwargs)
```
Comment (jamie256) [inline python/whylogs/api/writer/whylabs.py:650]: Consider returning from the commit? Maybe we get status or a code or error message?

File: python/whylogs/api/writer/whylabs.py [hunk 8]
```diff
@@ -606,6 +612,33 @@ def _tag_custom_perf_metrics(self, view: Union[DatasetProfileView, SegmentedData
...
+            logger.error("Must end current transaction with commit_transaction() before starting another")
+            return
+
+        if kwargs.get("dataset_id") is not None:
+            self._dataset_id = kwargs.get("dataset_id")
+
+        client: TransactionsApi = self._get_or_create_transaction_client()
+        request = TransactionStartRequest(dataset_id=self._dataset_id)
+        result: LogTransactionMetadata = client.start_transaction(request, **kwargs)
+        self._transaction_id = result["transaction_id"]
```
Comment (jamie256) [inline python/whylogs/api/writer/whylabs.py:635]: Error handling? How do I know if this worked as a caller? Maybe return values.

File: python/whylogs/api/writer/whylabs.py [hunk 9]
```diff
@@ -679,10 +712,26 @@ def write(self, file: Writable, **kwargs: Any) -> Tuple[bool, str]:
...
-                profile_file=tmp_file,
-            )
+            if self._transaction_id is not None:
+                # TODO: maybe check it's not a reference profile?
+                region = os.getenv("WHYLABS_UPLOAD_REGION", None)
+                client: TransactionsApi = self._get_or_create_transaction_client()
+                request = TransactionLogRequest(
+                    dataset_timestamp=dataset_timestamp_epoch, segment_tags=[], region=region
+                )
+                result: AsyncLogResponse = client.log_transaction(self._transaction_id, request, **kwargs)
```
Comment (jamie256) [inline python/whylogs/api/writer/whylabs.py:733]: consider info/debug log output for the result?

**Review by jamie256 — DISMISSED** [review]
Left a few comments, questions and suggestions. Can do now or in follow up.

File: python/whylogs/api/writer/whylabs.py (hunk 4 above)
Comment (richard-rogers) [reply python/whylogs/api/writer/whylabs.py:715]: Verified it throws on failure.

File: python/whylogs/api/writer/whylabs.py (hunk 7 above)
Comment (richard-rogers) [reply python/whylogs/api/writer/whylabs.py:650]: Verified if throws on failure. There doesn't seem to be anything interesting to return...

File: python/whylogs/api/writer/whylabs.py (hunk 8 above)
Comment (richard-rogers) [reply python/whylogs/api/writer/whylabs.py:635]: Verified it throws on failure.

**Review by jamie256 — APPROVED** [review]
LGTM!

**Pull request merged**